CMakeLists.txt

```diff
@@ -0,0 +1,60 @@
+cmake_minimum_required(VERSION 3.14 FATAL_ERROR)
+
+project(
+  CPM
+  VERSION 0.37.0
+  LANGUAGES NONE
+)
+
+option(CPM_INSTALL "Install CPM" OFF)
+
+include(CMakePackageConfigHelpers)
+configure_package_config_file(
+  "${CMAKE_CURRENT_SOURCE_DIR}/cmake/CPMConfig.cmake.in"
+  "${CMAKE_CURRENT_BINARY_DIR}/CPMConfig.cmake"
+  INSTALL_DESTINATION "${CMAKE_INSTALL_LIBDIR}/cmake/CPM"
+)
+write_basic_package_version_file(
+  "${CMAKE_CURRENT_BINARY_DIR}/CPMConfigVersion.cmake" COMPATIBILITY SameMajorVersion
+  ARCH_INDEPENDENT
+)
+
+# For consumer using FetchContent
+if(NOT ${CMAKE_PROJECT_NAME} STREQUAL ${PROJECT_NAME})
+  # Trick to use the find_package
+  configure_file(
+    "${CMAKE_CURRENT_SOURCE_DIR}/cmake/CPM.cmake" "${CMAKE_CURRENT_BINARY_DIR}/CPM.cmake" COPYONLY
+  )
+  set(CPM_DIR "${CMAKE_CURRENT_BINARY_DIR}")
+  # Check if the user download the repo with git and not in an official release
+  if(EXISTS "${CMAKE_CURRENT_SOURCE_DIR}/.git" AND IS_DIRECTORY "${CMAKE_CURRENT_SOURCE_DIR}/.git")
+    find_package(Git REQUIRED QUIET)
+    # Generate a git-describe version string from Git repository tags.
+    execute_process(
+      COMMAND ${GIT_EXECUTABLE} tag --points-at HEAD
+      WORKING_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}
+      OUTPUT_VARIABLE CPM_TAG
+      RESULT_VARIABLE GIT_DESCRIBE_ERROR_CODE
+      OUTPUT_STRIP_TRAILING_WHITESPACE
+    )
+    if(NOT GIT_DESCRIBE_ERROR_CODE AND NOT CPM_TAG STREQUAL "")
+      set(CPM_RELEASE TRUE)
+    else()
+      set(CPM_RELEASE FALSE)
+    endif()
+  endif()
+  find_package(CPM CONFIG REQUIRED NO_POLICY_SCOPE)
```

Comment:
How does this work? I've never thought you could `find_package()` itself especially without the `CPMConfig` files being generated. Did you test it on a clean machine?

Reply:
The CPMConfig is generated but not installed automatically. I think generally maybe it could not work but in our case we just need to include CPM.cmake so it works. It is working because I have provided CPM_DIR to say where to find the generated Config file

I did some test on my computer. It seems to work

Reply:
It seems you can use an Configfile generated inside the same project and CMake propose :
```
New in version 3.1: If the INSTALL_PREFIX argument is passed, this is used as base path to calculate all the relative paths. The <path> argument must be an absolute path. If this argument is not passed, the [CMAKE_INSTALL_PREFIX](https://cmake.org/cmake/help/latest/variable/CMAKE_INSTALL_PREFIX.html#variable:CMAKE_INSTALL_PREFIX) variable will be used instead. The default value is good when generating a FooConfig.cmake file to use your package from the install tree. When generating a FooConfig.cmake file to use your package from the build tree this option should be used.
```
but I don't really catch the meaning and how to use it.

Reply:
Ok, I see how that would work. But issue is that if you FetchContent against a Release tar ball, it will show as CPM_RELEASE=FALSE.

I think it would be more manageable to create a branch for each version where we set `CPM_RELEASE=TRUE` and on `main` we set it to `FALSE`. Then to create a new release we just branch main, rebase this commit, and tag release there.

Reply:
Ok, I see how that works, but this will mark release tarballs as `CPM_RELEASE=FALSE`.

A more manageable approach is to set `CPM_RELEASE=FALSE` on `main` branch. Then to publish a release:
- Branch `main` to `X.Y.Z` branch
- Rebase a commit that edits `set(CPM_RELEASE OFF)` -> `set(CPM_RELEASE ON)`
- Tag release to this branch

Reply:
To my understanding release tarballs will not have a .git folder inside so the CPM_RELEASE will be unset in this case. I see you point, we should check if CPM_RELEASE is define and then check if CPM_RELEASE is ON. What you propose would work but this requires CPM maintainers to agree on this behaviour. I tried to stay conservative on the workflows and CI

I think for now this would work :
```cmake
if(EXISTS "${CMAKE_CURRENT_SOURCE_DIR}/.git" AND IS_DIRECTORY "${CMAKE_CURRENT_SOURCE_DIR}/.git")
    find_package(Git REQUIRED QUIET)
    # Generate a git-describe version string from Git repository tags.
    execute_process(
      COMMAND ${GIT_EXECUTABLE} tag --points-at HEAD
      WORKING_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}
      OUTPUT_VARIABLE CPM_TAG
      RESULT_VARIABLE GIT_DESCRIBE_ERROR_CODE
      OUTPUT_STRIP_TRAILING_WHITESPACE
    )
    if(NOT GIT_DESCRIBE_ERROR_CODE AND NOT CPM_TAG STREQUAL "")
      set(CPM_RELEASE TRUE)
    else()
      set(CPM_RELEASE FALSE)
    endif()
else()
 set(CPM_RELEASE TRUE)
endif()
```

Reply:
Yes, there is also a possibility of the opposite (downloading a tarball for a specific commit), but let's ignore this for now.

I think we should ping the maintainers about the long-term approach, either here or a separate issue.

Reply:
I didn't know you could do this :) but it seems very a niche scenario, that it is not possible for now with get_cpm anyway.